### PR TITLE
Implement scheduler timers using asio deadline timers

### LINKED ISSUE #2218: Investigate the use of ASIO timers instead of the Scheduler tasks
The current solution of implementing timers (Scheduler tasks) with thread sleeps is likely to be [inefficient](https://gha.st/short-sleeps/) in terms of OS<->Userspace context switches and thread rescheduling for short sleep times (1 ms and less). We should investigate using the ASIO timer library (which uses the ```timerfd``` API) instead.

## Requirements:
- use one of ASIO's timer implementations instead of the Scheduler
- the timers should post a functor to the Dispatcher's execution queue
- should be safe when #2217 is resolved

### PR DESCRIPTION
**Pull request by ranisalt**
As it's been a while that #3056 was last updated, and there is usage of deprecated facilities. This is a bit cleaner using smart pointers to avoid manual deletion and skip the cache as constructing a timer seems very lightweight.

Closes #2218
Closes #3056

### CODE REVIEW

File: src/scheduler.cpp [hunk 1]
```diff
@@ -20,116 +20,69 @@
...
+		task->setEventId(++lastEventId);
 	}
 
 	// insert the event id in the list of active events
-	uint32_t eventId = task->getEventId();
-	eventIds.insert(eventId);
-
-	// add the event to the queue
-	eventList.push(task);
+	boost::asio::post(io_context, [this, task]() {
```
Comment (djarek) [inline src/scheduler.cpp:39]: Do we require C++14? If yes, `task` should be owned by a `unique_ptr` in the capture.

File: src/scheduler.cpp [hunk 2]
```diff
@@ -20,116 +20,69 @@
...
+		auto& timer = eventIdTimerMap[task->getEventId()];
+		timer.reset(new boost::asio::deadline_timer(io_context));
```
Comment (djarek) [inline src/scheduler.cpp:41]: I don't think there's a reason for the indirection in the mapped type, is there?

Comment (ranisalt) [reply src/scheduler.cpp:41]: @djarek at first I didn't think that a task and an entry in the timer map were 1:1, but I just noticed they are and I can cleanup this a bit.

File: src/scheduler.cpp (hunk 1 above)
Comment (ranisalt) [reply src/scheduler.cpp:39]: We don't :(

**Comment by yamaken93** [comment]
@ranisalt looks like a clean implementation. Is there is any tests to make sure it works like the previous one?

**Comment by ranisalt** [comment]
> Is there is any tests to make sure it works like the previous one?

I don't know for sure, but a couple of `addEvent`s to a player logging in should be enough to see.

**Review by MillhioreBT — APPROVED** [review]
Tested and it works, it also increases performance!

<details>
<summary>Test Code</summary>

```lua
GLOBAL_TEST_EV = {}

local talkTest = TalkAction("/tt")
function talkTest.onSay(player, words, param)
	if not player:getGroup():getAccess() then
		return true
	end

	if param == "test" then
		table.sort(GLOBAL_TEST_EV, function(a, b) return a < b end)
		print((GLOBAL_TEST_EV[#GLOBAL_TEST_EV] - GLOBAL_TEST_EV[1]) / 1000)
		GLOBAL_TEST_EV = {}
		return false
	end
	
	for i = 1, 10000 do
		addEvent(function(i)
			local s = ""
			for i = 1, 1000 do
				s = s .. 'i'
			end
			GLOBAL_TEST_EV[i] = os.mtime()
		end, 1000, i)
	end

	return false
end

talkTest:separator(" ")
talkTest:register()
```

</details>

After:
Screenshot_5
Before:
Screenshot_6

The improvement is evident, it takes less time for the processor to work on all scheduled tasks with these changes.

**Comment by ranisalt** [comment]
> The improvement is evident, it takes less time for the processor to work on all scheduled tasks with these changes.

:pray: hopefully this helps with high intensity guild wars on large servers.

**Comment by MillhioreBT** [comment]
Hi @SaiyansKing what is the correct test to verify that these changes are better than the previous code?
btw I would like to see your implementation in your optimized TFS

**Review by DSpeichert — DISMISSED** [review]

**Review by DSpeichert — APPROVED** [review]

**Review by MillhioreBT — APPROVED** [review]

**Comment by SaiyansKing** [comment]
> Hi @SaiyansKing what is the correct test to verify that these changes are better than the previous code?
I don't know what you mean by "better" but the thing is you expect it to give you free performance and that isn't the point of using asio deadline timers, the performance difference you get is because someone decided that it is better to put std::deque to std::priority_queue than std:;vector but std::vector performs much better(even boost asio use vector as underlying container for priority_queue) so if you would compare those changes on older tfs versions that used std::vector for std::priority_queue you wouldn't get a bit of performance gain.

> btw I would like to see your implementation in your optimized TFS
I don't use scheduler class at all since it is useless so it'll be difficult xD

**Comment by ranisalt** [comment]
Yes, this change is not a performance improvement, it's a code cleanup by leveraging a library with dedicated code that we already include.

As @SaiyansKing said, for future improvements, a lot of bottlenecks are related to poor choice of data structures. In general, we should avoid a list when a deque could be used, and avoid deques and forward_list when a vector could be used.

**Comment by gesior** [comment]
@ranisalt 
2 hosters reported to me random crashes on newest TFS. One sent backtrace, but without debug info:
https://otland.net/threads/tfs-service-main-process-exited-code-killed-status-11-segv.274811/#post-2646548
Can it be related to these changes?

**Comment by ranisalt** [comment]
@gesior it could be, it appears on the stack trace, but this change by itself does not mess with pointers. Most likely they used addEvent with a pointer.

**Comment by yamaken93** [comment]
@ranisalt https://github.com/otland/forgottenserver/blob/master/src/scheduler.cpp#L34 there is a data race here with multiple threads. That is why in my original code(which you used as base for this pr) i did the timer creation inside a lambda that was only run by the scheduler thread and i made sure that both the containers that handled timers were only accessed by scheduler thread so it was thread safe.

**Comment by ranisalt** [comment]
@yamaken93 thanks for the heads-up, I didn't notice `addEvent` could be called from other threads too.

**Comment by MillhioreBT** [comment]
> > Hi @SaiyansKing what is the correct test to verify that these changes are better than the previous code?
> 
> I don't know what you mean by "better" but the thing is you expect it to give you free performance and that isn't the point of using asio deadline timers, the performance difference you get is because someone decided that it is better to put std::deque to std::priority_queue than std:;vector but std::vector performs much better(even boost asio use vector as underlying container for priority_queue) so if you would compare those changes on older tfs versions that used std::vector for std::priority_queue you wouldn't get a bit of performance gain.
> 
> > btw I would like to see your implementation in your optimized TFS
> 
> I don't use scheduler class at all since it is useless so it'll be difficult xD

You are absolutely right!
With the fixes proposed in #3325 you no longer get the performance I mentioned above, it continues to work with the same performance of a lifetime

**Pull request merged**